### PR TITLE
Feature/single stats and tighten code

### PR DESCRIPTION
**Pull request by byb121**
CI test results: https://casm.pages.internal.sanger.ac.uk/cit-github/crisprreadcounts/results/feature-test_python_code/summary/

### CODE REVIEW

File: crispr_read_counts/single_guide_count.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  @staticmethod
+  def get_plasmid_read_counts(plasmid_file: str):
+    plasmid, plasmid_name = {}, {}
+    with open(plasmid_file, 'r') as f:
+      for line_count, line in enumerate(f):
+        ele = line.strip().split('\t')
+        if len(ele) < 3:
+          sys.exit(error_msg(f'Plasmid count file line: {line_count + 1} does not have 3 columns, or the file uses execpted delimiter.'))
+        if PLASMID_COUNT_HEADER.match(line):
+          plasmid_name = ele[2]
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_count.py]: The `plasmid_name` was initialised as a dictionary but here it is assigned a string and may be overwritten multiple times.

File: crispr_read_counts/single_guide_count.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    return lib, targeted_genes
+
+  @staticmethod
+  def get_plasmid_read_counts(plasmid_file: str):
+    plasmid, plasmid_name = {}, {}
+    with open(plasmid_file, 'r') as f:
+      for line_count, line in enumerate(f):
+        ele = line.strip().split('\t')
+        if len(ele) < 3:
+          sys.exit(error_msg(f'Plasmid count file line: {line_count + 1} does not have 3 columns, or the file uses execpted delimiter.'))
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_count.py]: Typo ('execpted').

File: crispr_read_counts/single_guide_count.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  @staticmethod
+  def get_single_guide_library(lib_file: str):
+    lib = {}
+    targeted_genes = {}
+
+    with open(lib_file, 'r') as f:
+      for line_count, line in enumerate(f):
+        ele = line.strip().split(',')
+        if len(ele) < 3:
+          sys.exit(error_msg(f'Guide RNA library file line: {line_count + 1} does not have 3 columns, or the file uses execpted delimiter.'))
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_count.py]: Typo ('execpted').

File: crispr_read_counts/single_guide_merge.py [hunk 4]
```diff
@@ -40,7 +49,10 @@ def get_counts_results(in_f: TextIO, with_or_without_plasmid_counts: bool):
             sample_count = sample.get(id, 0) + int(data[2])
             sample[id] = sample_count
             targeted_genes[id] = data[1]
-            plasmid[id] = int(data[3])
+            if id in plasmid and int(data[3]) != plasmid[id]:
+              sys.exit(error_msg(f'Plasmid count of sgRNA: {id} is not consistent across input count files.'))
+            else:
+              plasmid[id] = int(data[3])
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_merge.py]: Could be cast once before the conditional block.

File: crispr_read_counts/single_guide_count.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+      if self.plas_name:
+        f.write('\t'.join(['sgRNA', 'gene', f'{self.sample_name}.sample', self.plas_name]) + '\n')
+        for sgrna_seq in (sorted(self.lib.keys())):
+          for sgrna_id in self.lib[sgrna_seq]:
+            count = self.sample_count.get(sgrna_id, 0)
+            plasmid_count = self.plasmid.get(sgrna_id, 0)
+            f.write('\t'.join([sgrna_id, self.targeted_genes[sgrna_id], str(count), str(plasmid_count)]) + '\n')
+      else:
+        f.write('\t'.join(['sgRNA', 'gene', f'{self.sample_name}.sample']) + '\n')
+        for sgrna_seq in (sorted(self.lib.keys())):
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_count.py]: Unnecessary pair of parentheses.

File: crispr_read_counts/single_guide_count.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+            f.write('\t'.join([sgrna_id, self.targeted_genes[sgrna_id], str(count)]) + '\n')
+
+    self.stats['zero_count_guides'] = zero_count_guides
+    self.stats['low_count_guides'] = low_count_guides
+    with open(out_stats, 'w') as out_s:
+      json.dump(self.stats, out_s)
+      out_s.write('\n')
+
+  def _write_output_without_stats(self):
+    with open(self.out_count, 'w') as f:
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_count.py]: Using the `csv` package could simplify the write operations, though I don't know what the performance implications would be (if any).

File: crispr_read_counts/single_guide_count.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+
+    samfile = self.open_cram_and_get_sample_name()
+    lib_seqs, lib_seq_size = self.get_lib_seq_dict_and_seq_length(reverse_complementing)
+
+    for read in samfile.fetch(until_eof=True):
+      # if the alignment is secondary, supplymentary or  vendor failed, skip it!
+      if read.flag & 2816:
+        continue
+
+      if reverse_complementing:
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_count.py]: Reverse slicing is a corner case that has annoyed me before, so I investigated a bit...

I believe the following could be easier to digest and possibly faster (to be thoroughly checked).

Before the loop:

```python
sl = (
    slice(trim, trim + lib_seq_size) if not reverse_complementing else
    slice(-trim - lib_seq_size, -trim) if trim > 0 else
    slice(-lib_seq_size, None)
)
```

In the loop:

```python
cram_seq = read.get_forward_sequence()[sl]
```

File: crispr_read_counts/single_guide_count.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+        continue
+
+      # if the alignment is vendor failed, skip it but count it!
+      if read.flag & 512:
+        total_reads += 1
+        vendor_failed_reads += 1
+        continue
+
+      total_reads += 1
+      if reverse_complementing:
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_count.py]: See comment in `_get_sgrna_library_counts_without_stats`.

File: crispr_read_counts/single_guide_count.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+      if matching_lib_seq:
+        mapped_to_guide_reads += 1
+        for grna_id in self.lib[matching_lib_seq]:
+          self.sample_count[grna_id] = self.sample_count.get(grna_id, 0) + 1
+
+    self.stats['total_reads'] = total_reads
+    self.stats['vendor_failed_reads'] = vendor_failed_reads
+    self.stats['mapped_to_guide_reads'] = mapped_to_guide_reads
+
+  def _get_sgrna_library_counts_without_stats(self, trim: int, reverse_complementing: bool):
```
Comment (lbarbon) [inline crispr_read_counts/single_guide_count.py]: Pieces of code could be shared with `_get_sgrna_library_counts_with_stats` by putting them into functions, but function calls may then impact the performance. It's maintainability vs. performance tradeoff that may require some benchmarking to properly evaluate.

File: crispr_read_counts/utils.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+      f = gzip.open(file, 'rt')
+    else:
+      f = open(file, 'r')
+    try:
+        yield f
+    finally:
+        f.close()
+
+
+def rev_compl(dna: str):
```
Comment (lbarbon) [inline crispr_read_counts/utils.py]: I'd suggest a faster option (here only supporting uppercase letters, as it is implied to be required by `DNA_PATTERN`, though I don't know whether that applies to all sequences `rev_compl` is used with), with execution time about 19-44% of this implementation depending on the length of the sequence:

```python
dna_complement_tr_table = str.maketrans('ACGT', 'TGCA')

def reverse_complement(seq: str) -> str:
    return seq[::-1].translate(dna_complement_tr_table)
```

The current version is slowed down primarily by the call to `reversed` and the repeated initialisation of the translation table.

File: crispr_read_counts/single_guide_count.py (hunk 7 above)
Comment (byb121) [reply crispr_read_counts/single_guide_count.py]: Wow, this is wonderful!

File: crispr_read_counts/single_guide_count.py (hunk 9 above)
Comment (byb121) [reply crispr_read_counts/single_guide_count.py]: The performance tradeoff depends on the input sequencing file size. I can see that 11 million reads can have about 2% process time increase if stats are counted.

I haven't seen much bigger files yet, so I think I can squeeze the two functions into one.

File: crispr_read_counts/single_guide_count.py (hunk 6 above)
Comment (byb121) [reply crispr_read_counts/single_guide_count.py]: Using CSV does not have obvious performance implications, actually using CSV is marginally quicker.

Comment (byb121) [reply crispr_read_counts/single_guide_count.py]: When benchmarking dual_count module, using csv increases run time about 1.5 times when writing out 23 million lines.

Comment (lbarbon) [reply crispr_read_counts/single_guide_count.py]: I wonder if setting a [large] buffer size when opening the output file could improve things for `csv` methods (as well as raw writes). Never tried this, though.

Comment (byb121) [reply crispr_read_counts/single_guide_count.py]: Tried 100MB buffer when opening a file to write. It made little difference in performance for csv or raw writes...

**Pull request merged**